packages/xo-server-sdn-controller/src/index.js

```diff
@@ -406,124 +406,172 @@ class SDNController extends EventEmitter {
+          }
+
+          await privateNetwork.addNetwork(network)
+
+          // Previously created network didn't store `pif_device`
+          //
+          // 2019-08-22
+          // This is used to add the pif_device to networks created before this version. (v0.1.2)
+          // This will be removed in 1 year.
+          if (otherConfig['xo:sdn-controller:pif-device'] === undefined) {
+            const tunnel = getHostTunnelForNetwork(
+              privateNetwork.center,
+              network.$ref
+            )
+            const pif = xapi.getObjectByRef(tunnel.transport_PIF)
             await network.update_other_config(
-              'xo:sdn-controller:vni',
-              String(++this._prevVni)
+              'xo:sdn-controller:pif-device',
+              pif.device
             )
+          }
 
-            // Re-elect a center to apply the VNI
-            const privateNetwork = this._privateNetworks[
-              network.other_config['private-network-uuid']
-            ]
-            await this._electNewCenter(privateNetwork)
-          })
-        )
-      })
-    )
-  }
+          this._networks.set(network.$id, network.$ref)
+          if (privateNetwork.center !== undefined) {
+            this._starCenters.set(
+              privateNetwork.center.$id,
+              privateNetwork.center.$ref
+            )
+          }
+        })
+      )
 
-  async unload() {
-    this._privateNetworks = {}
-    this._newHosts = []
+      // Add VNI to other config of networks without VNI
+      //
+      // 2019-08-22
+      // This is used to add the VNI to networks created before this version. (v0.1.3)
+      // This will be removed in 1 year.
+      await Promise.all(
+        noVniNetworks.map(async network => {
+          await network.update_other_config(
+            'xo:sdn-controller:vni',
+            String(++this._prevVni)
+          )
 
-    this._networks.clear()
-    this._starCenters.clear()
+          // Re-elect a center to apply the VNI
+          const privateNetwork = this._privateNetworks[
+            network.other_config['private-network-uuid']
+          ]
+          await this._electNewCenter(privateNetwork)
+        })
+      )
+    } catch (error) {
+      log.error('Error while handling xapi connection', {
+        id: xapi.pool.uuid,
+        error,
+      })
+    }
+  }
 
-    this._cleaners.forEach(cleaner => cleaner())
-    this._cleaners = []
+  _handleDisconnectedXapi(xapi) {
+    log.debug('xapi disconnected', { id: xapi.pool.uuid })
+    try {
+      forOwn(this._privateNetworks, privateNetwork => {
+        privateNetwork.networks = omitBy(
+          privateNetwork.networks,
+          network => network.$pool.uuid === xapi.pool.uuid
+        )
 
-    this.ovsdbClients = {}
+        if (privateNetwork.center?.$pool.uuid === xapi.pool.uuid) {
+          this._electNewCenter(privateNetwork)
+        }
+      })
 
-    this._unsetApiMethods()
+      this._privateNetworks = filter(
```

Comment:
You can use native methods if you're not using any lodash's feature.

Reply:
It's pending in a different PR specifically to use less `lodash`